pqcheck/pqcheck.1.in
.TH PQCHECK 1 "2010-03-10"
.SH NAME
pqcheck - program to check that an LDM product queue was successfully closed
.SH SYNOPSIS
.HP
.ft B
pqcheck
.nh
\%[-F]
\%[-v]
\%[-l\ \fIlogdest\fP]
\%[-q\ \fIpqfname\fP]
.hy
.ft
.SH DESCRIPTION
.LP
This program checks that a product-queue is not open for writing by any
process.  Beginning with this version of the LDM, a product-queue contains
a persistent "write count".  Every time 
\fBpq_open\fP(3) is invoked on a product-queue with the
\fBPQ_READONLY\fP flag unset, the write-count is incremented.
Every time 
\fBpq_close\fP(3) is invoked on a product-queue opened in this way,
the write-count is decremented.
.LP
If it is known that no process holds the product-queue open for writing
and this utility indicates that the product-queue write-count is zero, then the
product-queue was successfully closed by all processes that had it open for
writing and the product-queue is guaranteed to be in a consistent state 
(i.e., the product-queue will not be corrupt).
.SH OPTIONS
.TP
.B -F
Force.  Support for a write-count will be added to the product-queue,
if necessary, and the write-count will be set to zero.
.TP
.BI \-l " logdest"
Log to \fIlogdest\fP. One of \fB''\fP (system logging daemon), \fB'-'\fP
(standard error stream), or file \fIlogdest\fP. Default is the standard error
stream if the process has a controlling terminal (i.e., the process isn't a
daemon); otherwise, either the LDM log file or the system logging daemon
(execute this program with just the option \fB'-?'\fP to determine which).
.TP
.BI "-q " pqfname
The filename of the product queue.
The default is
.nh
\fB$(regutil regpath{QUEUE_PATH})\fP.
.hy
.TP
.B -v
Verbose logging.  The write-count for the product-queue will be printed.
.SH SIGNALS
.TP
.BR SIGTERM
Normal termination.
.TP
.BR SIGINT
Immediate termination.

.SH EXIT STATUS
.TP
0
Success.  The product-queue was opened and the write-count is zero.
.TP
1
A system-error occurred.  
.TP
2
The product-queue doesn't have a write-count.  If you know that no process
has the product-queue open for writing, then you can use the \fB-F\fP option
to add write-count capability to the product-queue.
.TP
3
The product-queue was opened but the write-count is positive.
.TP
4
The product-queue could not be opened because it is internally inconsistent.
It will have to be deleted and recreated.

.SH EXAMPLE
.LP
The following shell-script fragment checks the product-queue:
.RS +4
.nf
#
# The product-queue exists.  Test for corruption.
#
if pqcheck -q $LDMHOME/data/ldm.pq; then
    echo "Using existing queue."
else
    case $? in
    1)  echo "Aborting..."
        ;;
    2)  echo "Adding write-count capability to product-queue"
        if ! pqcheck -F -q $LDMHOME/data/ldm.pq; then
            echo "Aborting..."
        fi
        ;;
    3)  echo "Product-queue incorrectly closed.  Checking..."
        if pqcat -s -l /dev/null; then
            echo "Product-queue appears OK.  Clearing write-count."
            if ! pqcheck -F -q $LDMHOME/data/ldm.pq; then
                echo "Aborting..."
            fi
        else
            echo "Product-queue appears corrupt.  Recreating..."
            /bin/mv -f $LDMHOME/data/ldm.pq $LDMHOME/data/ldm.pq.save
            /bin/su - ldm -c "$LDMBIN/ldmadmin delqueue"
            /bin/su - ldm -c "$LDMBIN/ldmadmin mkqueue"
        fi
        ;;
    4)  echo "Product-queue is corrupt.  Recreating..."
            /bin/mv -f $LDMHOME/data/ldm.pq $LDMHOME/data/ldm.pq.save
            /bin/su - ldm -c "$LDMBIN/ldmadmin delqueue"
            /bin/su - ldm -c "$LDMBIN/ldmadmin mkqueue"
        fi
        ;;
    esac
fi
.fi
.RE
This fragment
is appropriate for a shell-script that starts the LDM system at boot-time.

.SH FILES
.LP

.SH "SEE ALSO"
.LP
.BR ldmd (1),
.BR pqcat(1),
.BR ulog (3),
.BR pq (3),
.BR syslogd (8),
WWW URL \fBhttp://www.unidata.ucar.edu/software/ldm\fP.

.SH SUPPORT
.LP
If you have problems with this program, then you should first examine the 
LDM email archive for similar problems and how they were solved.
The email archive is available via the following World Wide Web URL:
.sp
.RS
\fBhttp://www.unidata.ucar.edu/software/ldm\fP
.RE
.sp
If this does not suffice and your site is a member of the Unidata 
program, then send an inquiry via email -- together will all relevant 
information -- to
.sp
.RS
\fBsupport@unidata.ucar.edu\fP
.RE


CHANGE_LOG
6.13.0.0	2016-02-11T12:02:01-0700
    Logging:
        Re-wrote LDM logging because:
            1) It was the major impediment to correctly using the LDM;
            2) It was the major impediment to putting the LDM in a Docker
               container; and
            3) Some system logging daemons sometimes dropped LDM log messages.
        For backward-compatibility, the original logging component, "ulog",
        still exists and is unconditionally included in the LDM library,
        "libldm". Indeed, the original behavior of the LDM can be obtained by
        specifying the option "--with-ulog" to the configure(1) script. By
        default, however, the new logging component, named "slog" (for "simple
        logging"), will be built and used. While "slog" can be told to write to
        the system logging daemon, by default it will write directly to the LDM
        log file if the process doesn't have a controlling terminal (i.e., is a
        daemon) or to the standard error stream if the process does have a
        controlling terminal (i.e., is not a daemon). Like the original "ulog"
        component, "slog" also supports log file rotation via a
        crontab(1)-executed "ldmadmin newlog" command. See the installed HTML
        documentation "$HOME/share/doc/ldm/basics/logging-format.html" or the
        web page
        <http://www.unidata.ucar.edu/software/ldm/ldm-current/basics/logfile-format.html>
        for more information.

6.12.15.42	2016-01-12T10:40:24-0700
    Make test for C++11-compatible g++ more robust.

6.12.15.41	2016-01-12T09:28:01-0700
    Get "make distclean" to work on both Winters (with C++11 support) and Spock
    (without C++11 support).

6.12.15.40	2016-01-11T16:07:56-0700
    Removed mylog2ulog stuff from ulog makefile.
    
    Eliminated memory-leaks from components that read GEMPAK tables.

6.12.15.39	2015-11-16T16:04:41-0700
    mldm_sender_manager(3):
        Add "-y" and "-z" options to execvp(2) of mldm_sender(1).

    pqact.c:
        Added SIGHUP forwarding.

6.12.15.38	2015-11-03T13:07:13-0700
    ldmd.c:
        Corrected some Coverity-discovered lint.

6.12.15.37	2015-11-03T09:41:04-0700
    Increase resolution of printed product-creation timestamp from millisecond
    to microsecond.

6.12.15.36	2015-11-02T15:59:18-0700
    ldmadmin(1):
        Reduce overestimation of new queue capacities (i.e., data capacity,
        product capacity) and don't compute new capacities if the data is
        insufficient.
        
    logging:
        Added logging options for using microsecond resolution in timestamps and
        formatting them according to the ISO 8601 standard. Added respective
        options "-y" and "-z" to ldmd(1) and notifyme(1).

6.12.15.35	2015-10-28T15:29:59-0600
    Create mcast_test(1) for testing LDM-7 on GENI.

6.12.15.34	2015-10-26T16:41:04-0600
    PerProdNotifier:
        Fix setting of product-information by notify_of_bop().

6.12.15.33	2015-10-23T15:57:05-0600
    mldm_sender():
        Moved addition of product-index-to-product-signature entry to before the
        sending of the product because the receiving LDM-7 would sometimes
        request a data-product before the mapping was entered resulting in the
        product not being found.

6.12.15.32	2015-10-22T19:32:42-0600
    Improve receiver session-memory API.

6.12.15.31	2015-10-20T17:11:53-0600
    down7.c:
        Correct location of down7_incNumProds().

        Add mutex destruction to down7_free().

    mldm_sender:
        Add opening of product-queue in thread-safe mode to prevent corruption
        of the regions-in-use list by multiple thread access.

    pq.c:
        Add unlocking of mutex in pq_sequence() before calling `ifMatch()`
        to avoid deadlock in calling a foreign function.

        Add mutex destruction to freeing.

6.12.15.30	2015-10-19T16:23:06-0600
    noaaportIngester(1):
        Added "+000 001 008 010 Total precipitation" to GEMPAK tables.

    Multicast:
        Add feature to lock products until VCMTP sender releases them.

        Modify VCMTP port numbers to match WireShark plugin.

        Improve log messages.

    pq(3):
        Make pq_release() thread-safe.

6.12.15.29	2015-10-08T18:03:25-0600
    Add diagnostic print statements and modify up7-down7_test to try
    diagnosing product-queue corruption seen by Shawn chen.

6.12.15.28	2015-10-06T16:03:42-0600
    pq(3):
        Modified fb_get() to log fblk metrics when it fails to help diagnose
        multicast problem.

6.12.15.27	2015-10-06T12:01:45-0600
    pqutil(1):
        Removed minor lint revealed by Coverity Scan.

    pq(3):
        Use correct product size in pqe_insert() and pq_insertNoSig().

    mcast_lib(3):
        Adapt to new checks by pqe_insert().

6.12.15.26	2015-10-06T09:48:58-0600
    pqact(1):
        Correct the setting of the product-queue cursor based on the last
        successfully-processed data-product from the previous session.
        Before this, pqact(1) might miss a product due to a restart of the LDM.
        This was most apparent in processing NEXRAD-2 data-products. Thanks
        Ryan.

    Product-Queue (pq(3)):
        Refactored pqe_insert() for better error-handling and return status.

        Harmonized return code names by adding PQ_DUP and PQ_BIG and making
        PQUEUE_DUP and PQUEUE_BIG aliases.

6.12.15.25	2015-10-01T15:28:56-0600
    Remove ignoring of SIGCONT from mldm_sender_manager because it's not needed.

    Generalize TCP keep-alive in TcpSend.cpp.

6.12.15.24	2015-09-30T15:40:21-0600
    Reduce number of "make distcheck" to just one in CI-commit.sh.

6.12.15.23	2015-09-30T14:33:58-0600
    Replace "$@" with just $@ in deploy.sh.

6.12.15.22	2015-09-30T13:32:25-0600
    Replace ${configOpts} with "$@" in deploy.sh to accommodate argument with
    space.

6.12.15.21	2015-09-29T16:57:17-0600
    Corrected setting of configure(1) options in delivery/deploy.sh.

6.12.15.20	2015-09-29T15:03:45-0600

pqcheck/pqcheck.c
/*
 *   Copyright 1993, University Corporation for Atmospheric Research
 *   See ../COPYRIGHT file for copying and redistribution conditions.
 */
/* $Id: pqcheck.c,v 1.2.4.1.4.3.2.2 2004/11/09 18:01:56 steve Exp $ */

/* 
 *  Check a product-queue.
 */

#include <config.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <rpc/rpc.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <assert.h>
#include <regex.h>
#include "ldm.h"
#include "atofeedt.h"
#include "globals.h"
#include "ldmprint.h"
#include "log.h"
#include "pq.h"
#include "md5.h"

#ifdef NO_ATEXIT
#include "atexit.h"
#endif

/* default "one trip" */
#ifndef DEFAULT_INTERVAL
#define DEFAULT_INTERVAL 0
#endif

#ifndef DEFAULT_FEEDTYPE
#define DEFAULT_FEEDTYPE ANY
#endif

static const char *pqfname;

static void
usage(const char *av0) /*  id string */
{
        (void)fprintf(stderr,
"Usage: %s [options]\n\tOptions:\n", av0);
        (void)fprintf(stderr,
"\t-F           Force. Set the writer-counter to zero "
"(creating it if necessary).\n");
        (void)fprintf(stderr,
"\t-v           Verbose\n");
        (void)fprintf(stderr,
"\t-l dest      Log to `dest`. One of: \"\" (system logging daemon), \"-\"\n"
"\t             (standard error), or file `dest`. Default is \"%s\"\n",
                log_get_default_destination());
        (void)fprintf(stderr,
"\t-q pqfname   (default \"%s\")\n", getDefaultQueuePath());
        (void)fprintf(stderr,
"Output defaults to standard output\n");
        exit(1);
}


static void
cleanup(void)
{
        log_notice("Exiting");
        (void)log_fini();
}


/*
 * Set signal handling.
 */
static void
set_sigactions(void)
{
        struct sigaction sigact;

        sigemptyset(&sigact.sa_mask);
        sigact.sa_flags = 0;

        /* Ignore these */
        sigact.sa_handler = SIG_IGN;
        (void) sigaction(SIGPIPE, &sigact, NULL);
        (void) sigaction(SIGALRM, &sigact, NULL);
        (void) sigaction(SIGCHLD, &sigact, NULL);
}


/*
 * Returns:
 *      0       Success.  Write-count of product-queue is zero.
 *      1       System failure.  See error-message.
 *      2       Product-queue doesn't support a writer-counter.  Not possible
 *              if "-F" option used.
 *      3       Write-count of product-queue is greater than zero.  Not possible
 *              if "-F" option used.
 *      4       The product-queue is internally inconsistent.
 */
int main(int ac, char *av[])
{
        const char *progname = basename(av[0]);
        int status = 0;
        unsigned write_count;
        int force = 0;

        /*
         * Set up error logging.
         */
        (void)log_init(progname);

        {
            extern int opterr;
            extern char *optarg;
            int ch;

            opterr = 1;

            while ((ch = getopt(ac, av, "Fvxl:q:")) != EOF)
                    switch (ch) {
                    case 'F':
                            force = 1;
                            break;
                    case 'v':
                            if (!log_is_enabled_info)
                                (void)log_set_level(LOG_LEVEL_INFO);
                            break;
                    case 'x':
                            (void)log_set_level(LOG_LEVEL_DEBUG);
                            break;
                    case 'l':
                            (void)log_set_destination(optarg);
                            break;
                    case 'q':
                            setQueuePath(optarg);
                            break;
                    case '?':
                            usage(progname);
                            break;
                    }
        }

        pqfname = getQueuePath();

        log_notice("Starting Up (%d)", getpgrp());

        /*
         * register exit handler
         */
        if(atexit(cleanup) != 0)
        {
                log_syserr("atexit");
                return 1;
        }

        /*
         * set up signal handlers
         */
        set_sigactions();

        if (force) {
            /*
             * Add writer-counter capability to the file, if necessary, and set
             * the writer-counter to zero.
             */
            status = pq_clear_write_count(pqfname);
            if (status) {
                if (PQ_CORRUPT == status) {
                    log_error("The product-queue \"%s\" is inconsistent", pqfname);
                    return 4;
                }
                else {
                    log_error("pq_clear_write_count() failure: %s: %s",
                            pqfname, strerror(status));
                    return 1;
                }
            }
            write_count = 0;
        }
        else {
            /*
             * Get the writer-counter of the product-queue.
             */
            status = pq_get_write_count(pqfname, &write_count);
            if (status) {
                if (ENOSYS == status) {
                    log_error("Product-queue \"%s\" doesn't have a writer-counter",
                        pqfname);
                    return 2;
                }
                else if (PQ_CORRUPT == status) {
                    log_error("Product-queue \"%s\" is inconsistent", pqfname);
                    return 4;
                }
                else {
                    log_error("pq_get_write_count() failure: %s: %s",
                        pqfname, strerror(status));
                    return 1;
                }
            }
        }

        log_info("The writer-counter of the product-queue is %u", write_count);

        return write_count == 0 ? 0 : 3;
}
